pkg/pulumiyaml/analyser.go

```diff
@@ -1065,37 +1065,49 @@ func (tc *typeCache) typeConfig(r *Runner, node configNode) bool {
 		switch {
 		case v.Default != nil:
 			// We have a default, so the type is optional
-			typ = tc.exprs[v.Default]
+			typCurrent = tc.exprs[v.Default]
 			optional = true
 		case v.Type != nil:
 			ctype, ok := ctypes.Parse(v.Type.Value)
 			if ok {
-				typ = ctype.Schema()
+				typCurrent = ctype.Schema()
 			}
 		}
 	case configNodeEnv:
 		if n.Type != nil {
-			typ = n.Type.Schema()
+			typCurrent = n.Type.Schema()
 		}
 	}
 
-	typ = &schema.InputType{ElementType: typ}
+	typCurrent = &schema.InputType{ElementType: typCurrent}
 	if optional {
-		typ = &schema.OptionalType{ElementType: typ}
+		typCurrent = &schema.OptionalType{ElementType: typCurrent}
 	}
 	// check for incompatible types between config/ configuration
 	if typExisting, ok := tc.configuration[k]; ok {
-		if typExisting.String() != typ.String() {
+		if typMatch, ok := tc.isConfigEquivalent(typExisting, typCurrent); !ok {
 			ctx := r.newContext(node)
 			ctx.error(nil, fmt.Sprintf("config key %s cannot have conflicting types %v, %v",
-				k, codegen.UnwrapType(typExisting), codegen.UnwrapType(typ)))
+				k, codegen.UnwrapType(typExisting), codegen.UnwrapType(typCurrent)))
 			return false
+		} else {
+			typCurrent = typMatch
 		}
 	}
-	tc.configuration[k] = typ
+	tc.configuration[k] = typCurrent
 	return true
 }
 
+func (tc *typeCache) isConfigEquivalent(a, b schema.Type) (schema.Type, bool) {
```

Comment:
This method is no longer a test for type equivalence. Generally `is*` methods return `bool`. I read this as "unifyConfigType(a, b) -> (type, ok)".

If it doesn't depend on `tc`, we don't need it to be a method.